toolbox/githubctl/main.go

```diff
@@ -79,167 +80,97 @@ func fastForward(repo, baseBranch, refSHA *string) error {
-	if err != nil {
-		return err
-	}
-	kv := processIstioVersion(&istioVersion)
-	pickledDeps, err := githubClnt.GetFileContent(istioRepo, *baseBranch, istioDepsFile)
-	if err != nil {
-		return err
-	}
-	deps, err := u.DeserializeDepsFromString(pickledDeps)
-	if err != nil {
-		return err
-	}
-	log.Printf("Fetching release tag\n")
-	releaseTag, err := getReleaseTag()
-	if err != nil {
-		return err
-	}
-	releaseMsg := "Istio Release " + releaseTag
-	for _, dep := range deps {
-		// use sha directly read from istio.VERSION in case updateVersion.sh was run
-		// manually without also updating istio.deps
-		log.Printf("Creating annotated tag [%s] on %s\n", releaseTag, dep.RepoName)
-		ref, exists := kv[dep.Name]
-		if !exists {
-			return fmt.Errorf("ill-defined %s: unable to find %s", istioVersionFile, dep.Name)
-		}
-		// make sure ref is a SHA, special case where previous release is used in this release
-		if u.ReleaseTagRegex.MatchString(ref) {
-			ref, err = githubClnt.GetTagCommitSHA(dep.RepoName, ref)
-			if err != nil {
-				return err
-			}
-		}
-		if err := githubClnt.CreateAnnotatedTag(
-			dep.RepoName, releaseTag, ref, releaseMsg); err != nil {
-			if strings.Contains(err.Error(), "Reference already exists") {
-				log.Printf("Tag [%s] already exists on %s\n", releaseTag, dep.RepoName)
-				prevTagSHA, err := githubClnt.GetTagCommitSHA(dep.RepoName, releaseTag)
-				if err != nil {
-					return err
+// preprocessProwResults downloads the most recent prow results up to maxRunDepth
+// then returns a two-level map job -> sha -> passed (true) or failed (false)
+func preprocessProwResults() map[string]map[string]bool {
+	log.Printf("Start preprocessing prow results")
+	prowAccessor := s.NewProwAccessor(
+		prowProject,
+		prowZone,
+		gubernatorURL,
+		gcsBucket,
+		u.NewGCSClient(gcsBucket))
+	cache := make(map[string]map[string]bool)
+	tasks := make(chan task)
+	var wg sync.WaitGroup
+	mutex := &sync.Mutex{}
+	for i := 0; i < *maxConcurrentRequests; i++ {
+		wg.Add(1)
+		go func() {
+			for {
+				t, more := <-tasks
+				if !more {
+					break
 				}
-				if prevTagSHA == ref {
-					log.Printf("Intended to tag [%s] at the same SHA, resort to no-op and continue\n", ref)
+				result, err := prowAccessor.GetResult(t.job, t.runNumber)
+				if err != nil {
+					log.Printf("failed to get result of %s at run number %d. Skip.", t.job, t.runNumber)
 					continue
-				} else {
-					return fmt.Errorf("trying to tag [%s] at different SHA", ref)
 				}
+				mutex.Lock()
+				cache[t.job][result.SHA] = result.Passed
+				mutex.Unlock()
 			}
-		}
-	}
-	return nil
-}
-
-// UpdateIstioVersionAfterReleaseTagsMadeOnDeps runs updateVersion.sh to update
-// istio.VERSION and then create a PR on istio
-func UpdateIstioVersionAfterReleaseTagsMadeOnDeps() error {
-	u.AssertNotEmpty("base_branch", baseBranch)
-	releaseTag, err := getReleaseTag()
-	if err != nil {
-		return err
+			wg.Done()
```

Comment:
move this to just before the for loop start and make it

defer wg.Done()

Reply:
done